File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.cpp
class Solution {
private:
    int m=-1;
    int n=-1;
    bool check(int x,int y){
        if(x<=-1 or x>=m or y<=-1 or y>=n){
              return false; 
        }
        return true;    
    }
    
public:
    int memo(int x, int y, vector<vector<int>>& matrix, vector<vector<int>>& dp){
        
        if(check(x,y) and dp[x][y]!=-1){
            return dp[x][y];
        }
        
        int val=matrix[x][y];
        vector<int> xs={0,-1,0,+1};
        vector<int> ys={-1,0,+1,0};
        m=matrix.size();
        n=matrix[0].size();
        
        int maxVal=1;
        for(int cnt=0;cnt<4;cnt++){
            int ptx=x+xs[cnt];
            int pty=y+ys[cnt];
            if(check(ptx,pty) and matrix[ptx][pty]>val){
                maxVal=max(maxVal,1+memo(ptx,pty,matrix, dp));
                //FML - wasted 3 hours here did 
                // maxVal=1+max(maxVal,foo(ptx,pty,matrix));
            }
        }
        return dp[x][y]=maxVal;
    }
    
        //135/139
    int foo(int x, int y, vector<vector<int>>& matrix){
        int val=matrix[x][y];
        vector<int> xs={0,-1,0,+1};
        vector<int> ys={-1,0,+1,0};
        m=matrix.size();
        n=matrix[0].size();
        
        int maxVal=1;
        for(int cnt=0;cnt<4;cnt++){
            int ptx=x+xs[cnt];
            int pty=y+ys[cnt];
            if(check(ptx,pty) and matrix[ptx][pty]>val){
                maxVal=max(maxVal,1+foo(ptx,pty,matrix));
                //FML - wasted 3 hours here did 
                // maxVal=1+max(maxVal,foo(ptx,pty,matrix));
            }
        }
        return maxVal;
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        vector<pair<int,int>> track;
        int ans=1;
        int m=matrix.size();
        int n=matrix[0].size();
        vector<vector<int>> dp(m,vector<int>(n,-1)); 
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                cout<<i<<" "<<j<<"\n";
                // ans=max(ans,foo(i,j, matrix));
                ans=max(ans,memo(i,j,matrix,dp));
            }
        }
        return ans;
    }
};

File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.py
class Solution:
    def __init__(self):
        self.m = -1
        self.n = -1

    def check(self, x, y):
        return 0 <= x < self.m and 0 <= y < self.n

    def memo(self, x, y, matrix, dp):
        if self.check(x, y) and dp[x][y] != -1:
            return dp[x][y]

        val = matrix[x][y]
        xs = [0, -1, 0, +1]
        ys = [-1, 0, +1, 0]
        self.m = len(matrix)
        self.n = len(matrix[0])

        maxVal = 1
        for cnt in range(4):
            ptx, pty = x + xs[cnt], y + ys[cnt]
            if self.check(ptx, pty) and matrix[ptx][pty] > val:
                maxVal = max(maxVal, 1 + self.memo(ptx, pty, matrix, dp))

        dp[x][y] = maxVal
        return maxVal

    def foo(self, x, y, matrix):
        val = matrix[x][y]
        xs = [0, -1, 0, +1]
        ys = [-1, 0, +1, 0]
        self.m = len(matrix)
        self.n = len(matrix[0])

        maxVal = 1
        for cnt in range(4):
            ptx, pty = x + xs[cnt], y + ys[cnt]
            if self.check(ptx, pty) and matrix[ptx][pty] > val:
                maxVal = max(maxVal, 1 + self.foo(ptx, pty, matrix))

        return maxVal

    def longestIncreasingPath(self, matrix):
        track = []
        ans = 1
        m = len(matrix)
        n = len(matrix[0])
        dp = [[-1 for _ in range(n)] for _ in range(m)]

        for i in range(m):
            for j in range(n):
                print(i, j)
                ans = max(ans, self.memo(i, j, matrix, dp))

        return ans
